lldb/source/Plugins/TypeSystem/Swift/LRUCache.h

```diff
@@ -0,0 +1,119 @@
+//===-- LRUCache.h ----------------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef liblldb_LRUCache_h_
+#define liblldb_LRUCache_h_
+
+#include "llvm/ADT/STLFunctionalExtras.h"
+#include "llvm/ADT/StringMap.h"
+#include <list>
+
+namespace lldb_private {
+namespace swift_demangle {
+
+template <typename Value>
+class LRUCache {
+public:
+  LRUCache(size_t capacity) : m_capacity(capacity) {}
+
+  std::optional<Value> Get(llvm::StringRef key) {
+    std::lock_guard lock{m_mutex};
+    auto map_it = m_map.find(key);
+    if (map_it == m_map.end())
+      return std::nullopt;
+    const auto &map_value = map_it->second;
+    MoveListElementToFront(map_value.list_it);
+    return map_value.value;
+  }
+
+  void Put(llvm::StringRef key, const Value &value) {
+    if (m_capacity == 0)
+      return;
+    std::lock_guard lock{m_mutex};
+    auto map_it = m_map.find(key);
+    if (map_it != m_map.end()) {
+      auto &map_value = map_it->second;
+      map_value.value = value;
+      MoveListElementToFront(map_value.list_it);
+    } else {
+      InsertElement(key, value);
+    }
+  }
+
+  Value GetOrCreate(llvm::StringRef key, llvm::function_ref<Value()> factory) {
+    if (m_capacity == 0)
+      return factory();
+
+    {
+      std::lock_guard lock{m_mutex};
+      if (auto map_it = m_map.find(key); map_it != m_map.end()) {
+        const auto &map_value = map_it->second;
+        MoveListElementToFront(map_value.list_it);
+        return map_value.value;
+      }
+    }
+
+    // Call `factory` with `m_mutex` unlocked
+    const auto value = factory();
+
+    {
+      std::lock_guard lock{m_mutex};
+      InsertElement(key, value);
+    }
+
+    return value;
+  }
+
+private:
+  using List = std::list<llvm::StringRef>;
```

Comment:
Would a vector<> and a moving "lru" iterator work and be faster?

Reply:
Can you elaborate, please? I'm not sure I understand how it could be faster. Both `Get` and `Put` are O(1) now. With a vector moving an element to the front would be O(N).

Reply:
I was thinking of storing the elements of the list in a vector and having an `lru` iterator that points to the newest entry. When inserting a new value we overwrite the value that is std::next(lru) and increment lru. And then we let that iterator wrap around like in a ring buffer. This would make scanning through the cache faster because we don't need to chase pointers and also use less memory. (Not that that matters for 10 elements). You are right that inserting an element would be more expensive because we need to copy up to N-1 elements. We could just always move the lru iterator to the right, and overwrite whatever is there whenever there is a cache miss. I think that should still give us a set of the last N used settings, we just don't know their relative order any more. But that might not matter.

Reply:
Got it now, thanks. Yeah, this can work, but we have to sacrifice reordering in `Get`. I think that's ok. It looks like most accesses to a cache item are sequential. In this case, pretty much any replacement strategy would work (even RR). I doubt it should be called "LRU" then as we effectively discard least recently _inserted_ items.
I'll prototype this variant.

Reply:
Oh that's a good point though. I didn't think o this being a global cache. We might be having a lot of long-lived types that we need to demangle over and over again. We might benefit from having a much larger cache, but then we probably need a DenseMap for fast lookups and maybe separately track the LRU list? I don't want to get hung up on the implementation details here. The nice thing is we can change and fine-tune it after an initial version landed, too. So from that point of view we can also keep your list-based implementation.

Reply:
`StringMap` copies the string key into the map, if you want to take advantage of the `ConstString` pool `DenseMap` is probably better.

Reply:
> Is DenseMap faster for lookups than StringMap?

No it's not. I think your current implementation is fine actually.

Reply:
But what Augusto said, is also true. The mangled names in the CompilerTypes are guaranteed to be ConstStrings, so a DenseMap<const char *, ...> is more efficient. We don't need to compare the strings at all, just the pointers.

Reply:
Yeah, you're probably right. The conversion `opaque_compiler_type_t->ConstString` will still do a lookup in a StringMap (one of the ConstString's pools), so it will be a `strlen` + 2-3 hash operations + the lookup itself. But we will store just pointers and not copy whole strings in the cache. So probably in the end, it will be the same performance with less memory footprint.